### DISCUSSION TITLE
React 狀態封裝

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#279] [前端相關問題] [answered]
最近一直在思考，react 裡面比較好的狀態管理方式到底應該要怎麼做？

典型的設計模式是：將狀態從父傳子，這樣的優點是團隊接手或 matain 相對容易；有些時候也會碰到要傳進去的狀態太多，就會將他要傳進去的 props 做分組，style 的是一組，他觸發的 function 會個別傳，就不會出現，以下這種狀況：你元件傳進去的 props 超爆賊多，團隊其他人要使用也相對不容易．

但這有一個缺點，很多時候我父層要宣告的狀態或是做的 call back (cb) 在不同頁面其實是一樣的，會出現大量的 copy paste 在做類似的事情，有沒有可能將這些東西也封裝好？比方說，我封裝好，不需要再父層每個頁面都做 onCancel & onOk 之類的操作

```js
// 傳進去沒有分組好，props 超爆賊多
const RemarkModal = ({
  visible,
  onOk,
  okText,
  onCancel,
  cancelText,
  title,
  remarkRequire,
  icon,
  placeholder,
  requiredRmk = true
}) => {
```
接著，我有看到這樣的做法，透過 hook return JSX component and its update function，我覺得這樣的好處是：狀態可以完好的封裝，只要透過 function 去 update 裡面的狀態就好了，超棒的吧！
但，對不是設計的人或是要一起協作的大家而言，會有三個 issue 
1. 封裝的考量不夠彈性，或是要擴充需求改變封裝的時候，要 trace 的比較深
2. 沒有文件的話，要先看過你的整個 hook 才會大概知道要怎麼使用操作與修改
3. 一些效能問題，每次 re-render 這個畫面就相當於是重新 new instance -> 所以每次的 component 就算長得一樣他都是截然不同的，所以要搭配 useMemo 的使用 -> 我不太確定這邊我的理解是否正確

老師或是大家會推薦怎麼樣的設計方式？會不推薦採用 hook 回傳 component 這種模式的設計嗎？ 

另外，想跟大家討論一下，大家的前端測試通常都會寫到什麼程度？

前端其實在測的是模仿使用者的行為看是否跟預期一樣，我目前最初步的都只有寫我打的 api 有沒有被 call 到，就醬ＸＤ 有點廢
但在寫 components 的測試時很容易不知不覺就變成整合測試，寫到非常詳細，但開發 >>>>>>>>>>>> 測試

想問問，通常大家實務上對測試會驗證到什麼樣的程度，或驗證哪些項目？

**Comment by aszx87410** [comment, marked as answer]
1. 以你舉的例子來說，我覺得其實傳進去的 props 並不會太多，數量滿合理的
2. 你會需要把狀態傳進去，就代表一定有其必要，所以這個省不了
3. 如果你每個頁面的 onCancel 跟 onOk 的行為都一樣，你可以用一個 hook 封裝起來，例如說 `const {visible, onCanecel, onOk} = useModal()` 之類的
4. 不推薦 hook 回傳 component，hook 比較適合處理邏輯相關的東西，你把一些 poprs 封裝在裡面就好了，像我上面那樣
5. 沒時間的話其實反而建議從 E2E 開始寫，因為跟底層實作比較無關，不用擔心實作細節改掉測試就要重寫，先從重要的邏輯開始測（登入登出之類的）

**Reply by s103071049** [reply]
謝謝老師ＸＤＤＤ

**Comment by CoolizzLuo** [comment]
props 你這個真的不算多XD，但有些 props 傳之前可以想一下是不是必須的。
例如大部分實作 modal，父層根本不需要知道 modal visible，Modal 也不需要知道裡面要搭載的元件，這時抽出來一個有 modal 功能的元件就可以比較彈性，這樣可以讓他可以任意搭配。

```javascript
const Modal = ({ onSubmit, onCancel, children }) => {
  const [isVisible, setIsVisible] = useState(false);
  const handleSubmit = () => {
    setIsVisible(false)
    onSubmit();
  }
  const handleCancel = () => {
    setIsVisible(false)
    onCancel();
  }

  return (
    <>
      <button onClick={() => setIsVisible(true)}>open</button>
      {isVisible &&
        (<div>
          {children}
          <div>
            <button onClick={handleSubmit}>submit</button>
            <button onClick={handleCancel}>cancel</button>
          </div>
        </div>)}
    </>
  )
}

const App = () => {
  return (
    <Modal onSubmit={() => {}} onCancel={() => {}}>
      <div>
        <h1>Modal Demo</h1>
        <i>icon</i>
        <p>blablabla</p>
      </div>
    </Modal>
  )
}
```

**Reply by aszx87410** [reply]
這個其實不太正確，像是 Ant Design 有 visible 屬性，MUI 也有 open 屬性，都是拿來控制 modal 是否顯示，而不是用「不顯示就不 render」這種方式
